ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveJoinWithFilterToAntiJoinRule.java

```diff
@@ -0,0 +1,145 @@
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.ql.optimizer.calcite.rules;
+
+import org.apache.calcite.plan.RelOptRule;
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelOptUtil;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Filter;
+import org.apache.calcite.rel.core.Join;
+import org.apache.calcite.rel.core.JoinRelType;
+import org.apache.calcite.rel.core.Project;
+import org.apache.calcite.rel.type.RelDataTypeField;
+import org.apache.calcite.rex.RexInputRef;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.util.ImmutableBitSet;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Planner rule that converts a join plus filter to anti join.
+ */
+public class HiveJoinWithFilterToAntiJoinRule extends RelOptRule {
+  protected static final Logger LOG = LoggerFactory.getLogger(HiveJoinWithFilterToAntiJoinRule.class);
+  public static final HiveJoinWithFilterToAntiJoinRule INSTANCE = new HiveJoinWithFilterToAntiJoinRule();
+
+  //    HiveProject(fld=[$0])
+  //      HiveFilter(condition=[IS NULL($1)])
+  //        HiveJoin(condition=[=($0, $1)], joinType=[left], algorithm=[none], cost=[not available])
+  //
+  // TO
+  //
+  //    HiveProject(fld_tbl=[$0])
+  //      HiveAntiJoin(condition=[=($0, $1)], joinType=[anti])
+  //
+  public HiveJoinWithFilterToAntiJoinRule() {
+    super(operand(Project.class, operand(Filter.class, operand(Join.class, RelOptRule.any()))),
+            "HiveJoinWithFilterToAntiJoinRule:filter");
+  }
+
+  // is null filter over a left join.
+  public void onMatch(final RelOptRuleCall call) {
+    final Project project = call.rel(0);
+    final Filter filter = call.rel(1);
+    final Join join = call.rel(2);
+    perform(call, project, filter, join);
+  }
+
+  protected void perform(RelOptRuleCall call, Project project, Filter filter, Join join) {
+    LOG.debug("Matched HiveAntiJoinRule");
+
+    assert (filter != null);
+
+    //We support conversion from left outer join only.
+    if (join.getJoinType() != JoinRelType.LEFT) {
+      return;
+    }
+
+    List<RexNode> aboveFilters = RelOptUtil.conjunctions(filter.getCondition());
+    boolean hasIsNull = false;
+
+    // Get all filter condition and check if any of them is a "is null" kind.
+    for (RexNode filterNode : aboveFilters) {
+      if (filterNode.getKind() == SqlKind.IS_NULL &&
+              isFilterFromRightSide(join, filterNode, join.getJoinType())) {
+        hasIsNull = true;
+        break;
+      }
+    }
+
+    // Is null should be on a key from right side of the join.
+    if (!hasIsNull) {
+      return;
+    }
+
+    // Build anti join with same left, right child and condition as original left outer join.
+    Join anti = join.copy(join.getTraitSet(), join.getCondition(),
```

Comment:
Probably it is here where we do not create the antijoin operator explicitly and why we end up with normal joins in Calcite plan. Since we are creating SemiJoin and AntiJoin as different operators, I think we should follow that pattern here and create an antijoin explicitely. Nevertheless, we could possibly get rid of HiveAntiJoin and HiveSemiJoin all together as I mentioned in another comment, but that can be part of another JIRA.

Reply:
done